File: app/src/main/java/com/beautifulquran/domain/EnglishTypography.kt
package com.beautifulquran.domain

/** Typographic punctuation policy for the punctuation-free English gloss. */
object EnglishTypography {
    private val terminalPunctuation = Regex("[.!?…][\\\"'’”)]*$")
    private val speechCue = Regex("(?:^|\\s)(say|says|said)$", RegexOption.IGNORE_CASE)
    private val nonSentenceCapital = Regex(
        "^(Allah|God|Lord|He|Him|His|We|Us|Our|I|Me|My|Musa|Firaun|" +
            "Prophet|Messenger|Most|Oft[- ]|All[- ])",
    )

    /**
     * Adds a stop before genuine capitalized sentence starts and after the
     * ayah's last gloss, without splitting on proper or reverential capitals.
     */
    fun punctuate(glosses: List<String>): List<String> = glosses.mapIndexed { index, gloss ->
        if (terminalPunctuation.containsMatchIn(gloss)) return@mapIndexed gloss
        val next = glosses.getOrNull(index + 1)?.trim().orEmpty()
        val capitalBoundary = next.firstOrNull()?.isUpperCase() == true &&
            !nonSentenceCapital.containsMatchIn(next) &&
            !speechCue.containsMatchIn(gloss.trim())
        if (index == glosses.lastIndex || capitalBoundary) "$gloss." else gloss
    }
}

File: web/src/domain/EnglishTypography.ts
const TERMINAL_PUNCTUATION = /[.!?…]["'’”)]*$/u
const SPEECH_CUE = /(?:^|\s)(?:say|says|said)$/iu
const NON_SENTENCE_CAPITAL = /^(?:Allah|God|Lord|He|Him|His|We|Us|Our|I|Me|My|Musa|Firaun|Prophet|Messenger|Most|Oft[- ]|All[- ])/u

/**
 * Adds restrained sentence stops to punctuation-free word-by-word glosses.
 * A capital after the first token starts a sentence unless it is a proper or
 * reverential capital, or follows a speech cue. The ayah's final token closes
 * with a stop so each rendered English paragraph has a visible ending.
 */
export function punctuateEnglishGlosses(glosses: readonly string[]): string[] {
  return glosses.map((gloss, index) => {
    if (TERMINAL_PUNCTUATION.test(gloss)) return gloss
    const next = glosses[index + 1]?.trim() ?? ''
    const capitalBoundary =
      next.length > 0 &&
      /^[A-Z]/u.test(next) &&
      !NON_SENTENCE_CAPITAL.test(next) &&
      !SPEECH_CUE.test(gloss.trim())
    return index === glosses.length - 1 || capitalBoundary ? `${gloss}.` : gloss
  })
}
